google-cloud-spanner/src/test/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClientMockServerTest.java

```diff
@@ -467,6 +478,188 @@ public void testWriteAtLeastOnceWithExcludeTxnFromChangeStreams() {
+    runner.run(
+        transaction -> {
+          try (ResultSet resultSet = transaction.executeQuery(STATEMENT)) {
+            while (resultSet.next()) {}
+          }
+
+          TransactionContextImpl impl = (TransactionContextImpl) transaction;
+          if (validTransactionId.get() == null) {
+            // Track the first not-null transactionId. This transaction gets ABORTED during commit
+            // operation and gets retried.
+            validTransactionId.set(impl.transactionId);
+          }
+          return null;
+        });
+
+    List<ExecuteSqlRequest> executeSqlRequests =
+        mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(2, executeSqlRequests.size());
+
+    // Verify the requests are executed using multiplexed sessions
+    for (ExecuteSqlRequest request : executeSqlRequests) {
+      assertTrue(mockSpanner.getSession(request.getSession()).getMultiplexed());
+    }
+
+    // Verify that the first request uses inline begin, and the previous transaction ID is set to
+    // ByteString.EMPTY
+    assertTrue(executeSqlRequests.get(0).hasTransaction());
+    assertTrue(executeSqlRequests.get(0).getTransaction().hasBegin());
+    assertTrue(executeSqlRequests.get(0).getTransaction().getBegin().hasReadWrite());
+    assertNotNull(
+        executeSqlRequests
+            .get(0)
+            .getTransaction()
+            .getBegin()
+            .getReadWrite()
+            .getMultiplexedSessionPreviousTransactionId());
+    assertEquals(
+        ByteString.EMPTY,
+        executeSqlRequests
+            .get(0)
+            .getTransaction()
+            .getBegin()
+            .getReadWrite()
+            .getMultiplexedSessionPreviousTransactionId());
+
+    // Verify that the second request uses inline begin, and the previous transaction ID is set
+    // appropriately
+    assertTrue(executeSqlRequests.get(1).hasTransaction());
+    assertTrue(executeSqlRequests.get(1).getTransaction().hasBegin());
+    assertTrue(executeSqlRequests.get(1).getTransaction().getBegin().hasReadWrite());
+    assertNotNull(
+        executeSqlRequests
+            .get(1)
+            .getTransaction()
+            .getBegin()
+            .getReadWrite()
+            .getMultiplexedSessionPreviousTransactionId());
+    assertNotEquals(
+        ByteString.EMPTY,
+        executeSqlRequests
+            .get(1)
+            .getTransaction()
+            .getBegin()
+            .getReadWrite()
+            .getMultiplexedSessionPreviousTransactionId());
+    assertEquals(
+        validTransactionId.get(),
+        executeSqlRequests
+            .get(1)
+            .getTransaction()
+            .getBegin()
+            .getReadWrite()
+            .getMultiplexedSessionPreviousTransactionId());
+  }
+*/
+  
+  // TODO(sriharshach): Uncomment test once Lock order preservation proto is published
+  /*
+  @Test
+  public void testAbortedReadWriteTxnUsesPreviousTxnIdOnRetryWithExplicitBegin()
```

Comment:
Test Scenario

1. Transaction 1 is started and the **inline begin** fails due to invalid update statement. This leads to ABORTED exception and transaction is retried.
2. Transaction 2 is initiated with **explicit begin** transaction which succeeds and creates a transactionId (**txn2**). however transaction2 gets aborted during commit (The test forces the commit rpc to return ABORTED the first time it is called).
3. Transaction 3 is started but again **inline begin** fails. Here txn id is null, but prevTransactionId will be txn2.
4. Transaction 4 is started with **explicit begin**. As Transaction 3 never got an ID, the transaction id of Transaction 2 (txn2) is used in TransactionOptions.
5. 
This behavior is verified in test by checking that second BeginTransaction RPC carries the **txn2** as previous transaction id.